### PR TITLE
feat(resharding): resharding archival node

### PR DESCRIPTION
**Pull request by staffik**
Mapping from child to parent shard is stored in `DBCol::StateShardUIdMapping`, which is marked as a cold column.
By default, it would be copied from hot to cold storage at each block.
In this PR, we do add `is_last_block_in_epoch` parameter to `update_cold_db()` so that it is only copied once per epoch.

For that, a slight refactor of `cold_store_copy()` was needed.
If it is possible to have skips at epoch boundary, then `cold_store_copy()` had a bug, where previous epoch shard layout could be used to copy data for a block from the new shard layout.

Added archival node and rpc node to reshardingv3 testloop.
In testloop, create temporary account before resharding and remove it after resharding. After gc period, the account is available at the archival node and it is not available at rpc node.
However, it works without actually testing the `cold_store_copy` path, because neither `cold_store_copy` nor garbage collection is called for archival node in testloop. So it needs fixing testloop and a follow-up testing.

### CODE REVIEW

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (staffik) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:165]: Yeah, just wanted to have at least one archival and one rpc so that they are tested. Not sure what bigger number of these would give us though.

File: core/store/src/archive/cold_storage.rs
Comment (wacban) [inline core/store/src/archive/cold_storage.rs:61]: If it is not too hard can you assert that this is the case?

File: core/store/src/archive/cold_storage.rs
Comment (wacban) [inline core/store/src/archive/cold_storage.rs:85]: What happened to the return value? Was is redundant?

File: core/store/src/archive/cold_storage.rs
Comment (wacban) [inline core/store/src/archive/cold_storage.rs:200]: nit: TODO(resharding) is the convention

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:183]: What about chunk validators?

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:942]: mini nit: Maybe move to separate variables?

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:945]: Is this sufficient to check that it returned ok?

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:1055]: 1) please add a comment what's this for
2) Do you want this to happen in all tests or should it be part of the TestParams?

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:1030]: Why not unpack? It was a bit nicer when unpacked.

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:1364]: Please avoid creating params twice, instead reuse the one that you have above. 

Some in other tests. 

nit: The following pattern may be a bit nicer:

```
let rpc_id = ...;
let params = TestReshardingParameters::new()
  .rpc_id(rpc_id)
  ...
```

File: nearcore/src/cold_storage.rs
Comment (wacban) [inline nearcore/src/cold_storage.rs:128]: That's fancy, this pattern is not commonly used. Would it be any more readable in a dedicated function?

**Review by wacban — APPROVED** [review]
LGTM, nice!

File: core/store/src/archive/cold_storage.rs
Comment (staffik) [reply core/store/src/archive/cold_storage.rs:61]: If it is not present, the function will return error.
Regarding finality, it was asserted in `cold_store_copy()`:
```
if next_height > hot_final_head_height {
            return Err(ColdStoreError::SkippedBlocksBetweenColdHeadAndNextHeightError {
                cold_head_height,
                next_height,
                hot_final_head_height,
            });
        }
```

File: core/store/src/archive/cold_storage.rs
Comment (staffik) [reply core/store/src/archive/cold_storage.rs:200]: I was using TODO(reshardingV3) for a few months to differentiate it from old todos for previous reshardingV2. Will replace all of them with TODO(resharding) now.

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (staffik) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:183]: Just refactored it a little bit in order to add archival node. Will add TODO to test chunk validators as well.

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (staffik) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:945]: Yes, if the result is `Result<QueryResponse, QueryError>`
and `QueryResponse` contains a data of a succesful response.

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (staffik) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:1055]: Added a comment. I would prefer it to happen in all tests, because it currently fails in a test that I did not expect to, so I think it can catch nontrivial bugs.

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (staffik) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:1030]: I also do not like it, but it is required by `create_account` / `delete_account`

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (staffik) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:1364]: I wanted to avoid it, but `rpc_id` is calculated in `TestReshardingParameters::new()`.

Comment (staffik) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:1364]: Mistakenly called `new()` twice in two tests, fixed it now.
But in general, I need:
```
    let params = TestReshardingParameters::new();
    let rpc_id = params.rpc_clients[0].clone();

    let params = params
        .add_loop_action(call_burn_gas_contract(
            ...
            rpc_id,
        ))
```
because `rpc_id` is calculated in `TestReshardingParameters::new()`.
Are you ok with that?

File: nearcore/src/cold_storage.rs
Comment (staffik) [reply nearcore/src/cold_storage.rs:128]: I do not see how make it more readable with a dedicated function.

Would this be good enough?
```
let next_height_block_hash = hot_store.get_ser::<CryptoHash>(DBCol::BlockHeight, &next_height.to_le_bytes())?;
if let Some(next_height_block_hash) = next_height_block_hash {
    break next_height_block_hash;
}
next_height = next_height + 1;
```

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (Trisfald) [reply integration-tests/src/test_loop/tests/resharding_v3.rs]: @staffik I proposed to make `TestReshardingParameters` use a builder pattern to delay construction at the end instead of initializing everything in `new` (#12610). Maybe it can make applying your changes easier.

Comment (staffik) [reply integration-tests/src/test_loop/tests/resharding_v3.rs]: That's amazing! Will wait for your PR to be merged in first.

File: nearcore/src/cold_storage.rs
Comment (wacban) [reply nearcore/src/cold_storage.rs]: Don't worry about it then, it's fine.

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (staffik) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:720]: Added TODO, won't do it in this PR.

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (staffik) [reply integration-tests/src/test_loop/tests/resharding_v3.rs]: Added TODO

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (staffik) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:586]: Without this shard shuffling test failed because default number of seats was 100, and it treated rpc and archival nodes as validators.

File: core/store/src/archive/cold_storage.rs
Comment (shreyan-gupta) [inline core/store/src/archive/cold_storage.rs:100]: Not a big fan of this `is_last_block_in_epoch` check here. Is it possible to just remove this and copy DBCol::StateShardUIdMapping at every height? I don't suppose it would add too much inefficiency, would it?

File: core/store/src/adapter/trie_store.rs
Comment (shreyan-gupta) [inline core/store/src/adapter/trie_store.rs:178]: Could we push this pub function to `TrieStoreAdapter`? In cold_storage.rs we can call `hot_store.trie_store().get_shard_uid_mapping(...)`

File: nearcore/src/cold_storage.rs
Comment (shreyan-gupta) [inline nearcore/src/cold_storage.rs:139]: Wait, this isn't correct right? Shouldn't we be calling `update_cold_db` function for all heights between `cold_head_height + 1` and `hot_final_head_height`? It should be within the loop right?

**Review by shreyan-gupta — COMMENTED** [review]
Could you please check my comment in nearcore/src/cold_storage.rs before merging?

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (Trisfald) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:586]: indeed.. I've run into the same issue while adding other node types.

File: core/store/src/adapter/trie_store.rs
Comment (staffik) [reply core/store/src/adapter/trie_store.rs:178]: I would like to but it is problematic, please see https://github.com/near/nearcore/pull/12232#discussion_r1804810508.

File: nearcore/src/cold_storage.rs
Comment (staffik) [reply nearcore/src/cold_storage.rs:139]: > copies data for the next available produced block after current cold store head

The loop exist in order to find the first available block in hot store

File: core/store/src/adapter/trie_store.rs
Comment (shreyan-gupta) [reply core/store/src/adapter/trie_store.rs:178]: Aah okay! Yes, we don't want to be cloning store I guess. Could you add a quick comment just so that next time we have context for why it's where it is?

File: nearcore/src/cold_storage.rs
Comment (shreyan-gupta) [reply nearcore/src/cold_storage.rs:139]: No but in the original code, `update_cold_db` is being called in the loop for each height right?? That has side effects??

Comment (shreyan-gupta) [reply nearcore/src/cold_storage.rs:139]: Ohhh..... Originally poorly written 😢
Apologies for the panic!()

File: core/store/src/adapter/trie_store.rs
Comment (staffik) [reply core/store/src/adapter/trie_store.rs:181]: Added here: f46714f3c903c73ed30610fa351bc74a06a96f0a

**Pull request merged**
